File: packages/qcraft-engine/src/qcraft_engine/climate.py
"""Climate scenario module (calc_climate_scenario).

Recomputes fiscal projections under climate-adjusted productivity growth.
Called once per scenario (6 times total). Uses explicit Python for-loops
for recursive GDP and debt dynamics (domain rule #1).

Key domain rules:
- NO debt floor: climate scenarios can produce negative debt-to-GDP (rule #3)
- Employment growth unchanged from baseline
- Interest rate and inflation unchanged from baseline
- No fiscal rule applied within climate scenarios
- Expenditure rigidity 1.0 = sticky (worst case), 0.0 = flexible (rule #4)
"""

import polars as pl

YEAR_START = 2009
YEAR_END = 2099


def calc_climate_scenario(
    data_baseline: pl.DataFrame,
    data_baseline_v1: pl.DataFrame,
    data_interest: pl.DataFrame,
    climate_variation: pl.DataFrame,
    expenditure_rigidity: float = 1.0,
    data_risk: pl.DataFrame | None = None,
) -> pl.DataFrame:
    """Compute climate-adjusted fiscal projections for one scenario.

    Args:
        data_baseline: Output of baseline_country() (fiscal module). Columns:
            years, revenue, revenue_percent_gdp, primary_expenditure,
            primary_expenditure_percent_gdp, debt_to_gdp, debt, etc.
        data_baseline_v1: Output of baseline_v1(). Columns:
            years, employment_growth, labour_productivity_growth,
            gdp_deflator_growth_percent, nominal_gdp, real_gdp,
            nominal_gdp_growth_percent.
        data_interest: Output of interest_rate_country(). Columns:
            years, nominal_interest_rate.
        climate_variation: DataFrame with columns: years, climate_variation.
            The year-over-year productivity growth shock. Zero for years
            <= WEO_MAX_YEAR, nonzero from climate impact start year.
        expenditure_rigidity: 0.0 (flexible) to 1.0 (sticky, default).
        data_risk: Optional discrete risks DataFrame. If provided, must have
            columns: years, revenue_risk, expenditure_risk (% GDP).

    Returns:
        DataFrame with 21 columns, years 2009-2099 (91 rows).
    """
    # Sort all inputs
    bv1 = data_baseline_v1.filter(
        pl.col("years").is_between(YEAR_START, YEAR_END)
    ).sort("years")
    fiscal = data_baseline.filter(
        pl.col("years").is_between(YEAR_START, YEAR_END)
    ).sort("years")
    interest = data_interest.filter(
        pl.col("years").is_between(YEAR_START, YEAR_END)
    ).sort("years")
    cv = climate_variation.sort("years")

    # Determine WEO_MAX_YEAR: last year where climate_variation is 0
    # (climate impacts start at WEO_MAX_YEAR + 1)
    cv_years = cv["years"].to_list()
    cv_vals = cv["climate_variation"].to_list()

    # Find WEO_MAX_YEAR from the data: the year before first nonzero variation
    weo_max_year = YEAR_START
    for y, v in zip(cv_years, cv_vals):
        if v != 0.0:
            weo_max_year = y - 1
            break
    else:
        # All zeros — no climate impact. Use last year before projection.
        weo_max_year = 2029

    # Build lookups from baseline_v1
    bv1_lookup: dict[int, dict[str, float]] = {}
    for row in bv1.iter_rows(named=True):
        y = int(row["years"])
        bv1_lookup[y] = {
            "employment_growth": float(row["employment_growth"]),
            "labour_productivity_growth": float(row["labour_productivity_growth"]),
            "gdp_deflator_growth_percent": float(row["gdp_deflator_growth_percent"]),
            "nominal_gdp": float(row["nominal_gdp"]),
            "real_gdp": float(row["real_gdp"]),
            "nominal_gdp_growth_percent": float(row["nominal_gdp_growth_percent"]),
        }

    # Build lookup from fiscal (baseline_country output)
    fiscal_lookup: dict[int, dict[str, float]] = {}
    for row in fiscal.iter_rows(named=True):
        y = int(row["years"])
        fiscal_lookup[y] = {
            "revenue": float(row["revenue"]),
            "revenue_percent_gdp": float(row["revenue_percent_gdp"]),
            "primary_expenditure": float(row["primary_expenditure"]),
            "primary_expenditure_percent_gdp": float(
                row["primary_expenditure_percent_gdp"]
            ),
            "primary_balance": float(row["primary_balance"]),
            "primary_balance_percent_gdp": float(row["primary_balance_percent_gdp"]),
            "interest_expenditure": float(row["interest_expenditure"]),
            "interest_expenditure_percent_gdp": float(
                row["interest_expenditure_percent_gdp"]
            ),
            "total_expenditure": float(row["total_expenditure"]),
            "overall_balance": float(row["overall_balance"]),
            "overall_balance_percent_gdp": float(row["overall_balance_percent_gdp"]),
            "debt_to_gdp": float(row["debt_to_gdp"]),
            "debt": float(row["debt"]),
        }

    # Build lookup from interest rate
    interest_lookup: dict[int, float] = {}
    for row in interest.iter_rows(named=True):
        interest_lookup[int(row["years"])] = float(row["nominal_interest_rate"])

    # Build climate variation lookup
    cv_lookup: dict[int, float] = {}
    for y, v in zip(cv_years, cv_vals):
        cv_lookup[int(y)] = float(v)

    # Build discrete risk lookups
    risk_rev_lookup: dict[int, float] = {}
    risk_exp_lookup: dict[int, float] = {}
    if data_risk is not None:
        for row in data_risk.iter_rows(named=True):
            y = int(row["years"])
            risk_rev_lookup[y] = float(row["revenue_risk"])
            risk_exp_lookup[y] = float(row["expenditure_risk"])

    # Output arrays
    years_out = list(range(YEAR_START, YEAR_END + 1))
    n = len(years_out)

    # GDP arrays
    labour_prod_growth = [0.0] * n
    real_gdp_growth = [0.0] * n
    nominal_gdp_growth = [0.0] * n
    nominal_gdp = [0.0] * n
    real_gdp = [0.0] * n
    employment_growth = [0.0] * n

    # Fiscal arrays
    revenue = [0.0] * n
    revenue_pct = [0.0] * n
    primary_exp = [0.0] * n
    primary_exp_pct = [0.0] * n
    primary_bal = [0.0] * n
    primary_bal_pct = [0.0] * n
    interest_exp = [0.0] * n
    interest_exp_pct = [0.0] * n
    total_exp = [0.0] * n
    overall_bal = [0.0] * n
    overall_bal_pct = [0.0] * n
    debt_to_gdp = [0.0] * n
    debt = [0.0] * n
    dspb: list[float | None] = [None] * n

    # Phase 1-2: Fill WEO period (years <= WEO_MAX_YEAR) from baseline
    for i, year in enumerate(years_out):
        if year > weo_max_year:
            break

        bv = bv1_lookup[year]
        fl = fiscal_lookup[year]

        # GDP columns from baseline_v1
        labour_prod_growth[i] = bv["labour_productivity_growth"]
        real_gdp_growth[i] = float(
            bv1.filter(pl.col("years") == year)["real_gdp_growth_percent"][0]
        )
        nominal_gdp_growth[i] = bv["nominal_gdp_growth_percent"]
        nominal_gdp[i] = bv["nominal_gdp"]
        real_gdp[i] = bv["real_gdp"]
        employment_growth[i] = bv["employment_growth"]

        # Fiscal columns from baseline_country
        revenue[i] = fl["revenue"]
        revenue_pct[i] = fl["revenue_percent_gdp"]
        primary_exp[i] = fl["primary_expenditure"]
        primary_exp_pct[i] = fl["primary_expenditure_percent_gdp"]
        primary_bal[i] = fl["primary_balance"]
        primary_bal_pct[i] = fl["primary_balance_percent_gdp"]
        interest_exp[i] = fl["interest_expenditure"]
        interest_exp_pct[i] = fl["interest_expenditure_percent_gdp"]
        total_exp[i] = fl["total_expenditure"]
        overall_bal[i] = fl["overall_balance"]
        overall_bal_pct[i] = fl["overall_balance_percent_gdp"]
        debt_to_gdp[i] = fl["debt_to_gdp"]
        debt[i] = fl["debt"]

        # DSPB (from 2010 onward)
        if i > 0:
            nom_rate = interest_lookup[year]
            gdp_g = nominal_gdp_growth[i]
            dspb[i] = debt_to_gdp[i - 1] * (nom_rate - gdp_g) / 100 / (1 + gdp_g / 100)

    # Phase 2-5: Recursive computation for projection years
    for i, year in enumerate(years_out):
        if year <= weo_max_year:
            continue

        bv = bv1_lookup[year]
        fl = fiscal_lookup[year]
        nom_rate = interest_lookup[year]
        variation = cv_lookup.get(year, 0.0)
        inflation = bv["gdp_deflator_growth_percent"]

        # Phase 1: Adjust productivity
        employment_growth[i] = bv["employment_growth"]
        labour_prod_growth[i] = bv["labour_productivity_growth"] + variation

        # Phase 2: Recompute GDP (recursive)
        real_gdp_growth[i] = (1 + employment_growth[i] / 100) * (
            1 + labour_prod_growth[i] / 100
        ) * 100 - 100
        nominal_gdp_growth[i] = (1 + real_gdp_growth[i] / 100) * (
            1 + inflation / 100
        ) * 100 - 100
        real_gdp[i] = real_gdp[i - 1] * (1 + real_gdp_growth[i] / 100)
        nominal_gdp[i] = nominal_gdp[i - 1] * (1 + nominal_gdp_growth[i] / 100)

        # Phase 3: Expenditure recalibration
        baseline_pexp = fl["primary_expenditure"]
        baseline_pexp_pct = fl["primary_expenditure_percent_gdp"]
        primary_exp_with_baseline_pct = baseline_pexp_pct * nominal_gdp[i] / 100
        recalibration = baseline_pexp - primary_exp_with_baseline_pct
        primary_exp[i] = baseline_pexp - (1 - expenditure_rigidity) * recalibration

        # Phase 4: Revenue (constant ratio from baseline)
        revenue_pct[i] = fl["revenue_percent_gdp"]
        revenue[i] = revenue_pct[i] / 100 * nominal_gdp[i]

        # Apply discrete risks if provided
        if data_risk is not None:
            rev_risk = risk_rev_lookup.get(year, 0.0)
            exp_risk = risk_exp_lookup.get(year, 0.0)
            revenue[i] += rev_risk / 100 * nominal_gdp[i]
            primary_exp[i] += exp_risk / 100 * nominal_gdp[i]

        # Phase 5: Recursive fiscal (NO debt floor)
        primary_exp_pct[i] = primary_exp[i] / nominal_gdp[i] * 100
        primary_bal[i] = revenue[i] - primary_exp[i]
        primary_bal_pct[i] = primary_bal[i] / nominal_gdp[i] * 100

        # Debt dynamics — NO max(0, ...)
        debt_to_gdp[i] = (
            debt_to_gdp[i - 1]
            * (1 + nom_rate / 100)
            / (1 + nominal_gdp_growth[i] / 100)
            - primary_bal_pct[i]
        )

        # Debt level
        debt[i] = debt_to_gdp[i] / 100 * nominal_gdp[i]

        # Interest expenditure (prior-year debt * current rate)
        interest_exp[i] = debt[i - 1] * nom_rate / 100
        interest_exp_pct[i] = interest_exp[i] / nominal_gdp[i] * 100

        # Total expenditure and overall balance
        total_exp[i] = primary_exp[i] + interest_exp[i]
        overall_bal[i] = revenue[i] - total_exp[i]
        overall_bal_pct[i] = overall_bal[i] / nominal_gdp[i] * 100

        # DSPB
        dspb[i] = (
            debt_to_gdp[i - 1]
            * (nom_rate - nominal_gdp_growth[i])
            / 100
            / (1 + nominal_gdp_growth[i] / 100)
        )

    return pl.DataFrame(
        {
            "years": years_out,
            "revenue": revenue,
            "revenue_percent_gdp": revenue_pct,
            "primary_expenditure": primary_exp,
            "primary_expenditure_percent_gdp": primary_exp_pct,
            "primary_balance": primary_bal,
            "primary_balance_percent_gdp": primary_bal_pct,
            "interest_expenditure": interest_exp,
            "interest_expenditure_percent_gdp": interest_exp_pct,
            "total_expenditure": total_exp,
            "overall_balance": overall_bal,
            "overall_balance_percent_gdp": overall_bal_pct,
            "debt_to_gdp": debt_to_gdp,
            "debt": debt,
            "debt_stabilizing_primary_balance": dspb,
            "labour_productivity_growth": labour_prod_growth,
            "real_gdp_growth_percent": real_gdp_growth,
            "nominal_gdp_growth_percent": nominal_gdp_growth,
            "nominal_gdp": nominal_gdp,
            "real_gdp": real_gdp,
            "employment_growth": employment_growth,
        }
    )

File: packages/qcraft-engine-ts/src/climate.ts
/**
 * Climate scenario module (`calc_climate_scenario` in Python).
 *
 * Recomputes the fiscal projection under climate-adjusted productivity growth. Called
 * once per scenario (six times per country). Recursive GDP and debt dynamics use an
 * explicit year-by-year loop (domain rule #1).
 *
 * Key domain rules encoded here:
 * - NO debt floor: climate scenarios may produce negative debt-to-GDP (rule #3).
 * - Employment growth, interest rate and inflation are unchanged from baseline.
 * - No fiscal rule feedback is applied inside a climate scenario.
 * - `expenditureRigidity` 1.0 = sticky (worst case), 0.0 = fully flexible (rule #4).
 */

import type {
  BaselineV1Row,
  ClimateRow,
  ClimateVariationRow,
  FiscalRow,
  InterestRateRow,
  Num,
  RiskRow,
} from './types.js';
import { YEAR_END, YEAR_START } from './constants.js';
import { mustGet } from './internal.js';

export interface ClimateOptions {
  /** Explicit first application year for the rolling profile; legacy callers retain inference. */
  climateStartYear?: number;
  /** 0.0 (flexible) to 1.0 (sticky, default). 1.0 keeps expenditure at baseline levels. */
  expenditureRigidity?: number;
  /** Optional discrete revenue/expenditure shocks, in % of GDP. */
  dataRisk?: readonly RiskRow[] | null;
}

/**
 * Compute climate-adjusted fiscal projections for one scenario.
 *
 * @param dataBaseline Output of `baselineCountry` (the fiscal module).
 * @param dataBaselineV1 Output of `baselineV1`.
 * @param dataInterest Output of `interestRateCountry`.
 * @param climateVariation Year-over-year productivity growth shock; zero through
 *   WEO_MAX_YEAR, which is how the function infers where projections begin.
 * @returns 91 rows (2009–2099) with 21 columns.
 */
export function calcClimateScenario(
  dataBaseline: readonly FiscalRow[],
  dataBaselineV1: readonly BaselineV1Row[],
  dataInterest: readonly InterestRateRow[],
  climateVariation: readonly ClimateVariationRow[],
  options: ClimateOptions = {},
): ClimateRow[] {
  const { expenditureRigidity = 1.0, dataRisk = null } = options;

  const inRange = <T extends { years: number }>(rows: readonly T[]): T[] =>
    rows
      .filter((r) => r.years >= YEAR_START && r.years <= YEAR_END)
      .sort((a, b) => a.years - b.years);

  const bv1 = inRange(dataBaselineV1);
  const fiscal = inRange(dataBaseline);
  const interest = inRange(dataInterest);
  const cv = [...climateVariation].sort((a, b) => a.years - b.years);

  // WEO_MAX_YEAR is the year before the first nonzero climate variation.
  const firstShock = cv.find((r) => r.climate_variation !== 0.0);
  const weoMaxYear = options.climateStartYear === undefined
    ? (firstShock === undefined ? 2029 : firstShock.years - 1)
    : options.climateStartYear - 1;

  const bv1Lookup = new Map<number, BaselineV1Row>();
  for (const row of bv1) bv1Lookup.set(row.years, row);

  const fiscalLookup = new Map<number, FiscalRow>();
  for (const row of fiscal) fiscalLookup.set(row.years, row);

  const interestLookup = new Map<number, number>();
  for (const row of interest) interestLookup.set(row.years, row.nominal_interest_rate);

  const cvLookup = new Map<number, number>();
  for (const row of cv) cvLookup.set(row.years, row.climate_variation);

  const riskRevLookup = new Map<number, number>();
  const riskExpLookup = new Map<number, number>();
  if (dataRisk) {
    for (const row of dataRisk) {
      riskRevLookup.set(row.years, row.revenue_risk);
      riskExpLookup.set(row.years, row.expenditure_risk);
    }
  }

  const years: number[] = [];
  for (let y = YEAR_START; y <= YEAR_END; y += 1) years.push(y);
  const n = years.length;

  // GDP arrays.
  const labourProdGrowth = new Array<number>(n).fill(0);
  const realGdpGrowth = new Array<number>(n).fill(0);
  const nominalGdpGrowth = new Array<number>(n).fill(0);
  const nominalGdp = new Array<number>(n).fill(0);
  const realGdp = new Array<number>(n).fill(0);
  const employmentGrowth = new Array<number>(n).fill(0);

  // Fiscal arrays.
  const revenue = new Array<number>(n).fill(0);
  const revenuePct = new Array<number>(n).fill(0);
  const primaryExp = new Array<number>(n).fill(0);
  const primaryExpPct = new Array<number>(n).fill(0);
  const primaryBal = new Array<number>(n).fill(0);
  const primaryBalPct = new Array<number>(n).fill(0);
  const interestExp = new Array<number>(n).fill(0);
  const interestExpPct = new Array<number>(n).fill(0);
  const totalExp = new Array<number>(n).fill(0);
  const overallBal = new Array<number>(n).fill(0);
  const overallBalPct = new Array<number>(n).fill(0);
  const debtToGdp = new Array<number>(n).fill(0);
  const debt = new Array<number>(n).fill(0);
  const dspb = new Array<Num>(n).fill(null);

  // --- WEO period: copy the baseline through unchanged ---
  for (let i = 0; i < n; i += 1) {
    const year = years[i]!;
    if (year > weoMaxYear) break;

    const bv = mustGet(bv1Lookup, year, 'baseline_v1');
    const fl = mustGet(fiscalLookup, year, 'fiscal baseline');

    labourProdGrowth[i] = bv.labour_productivity_growth;
    realGdpGrowth[i] = bv.real_gdp_growth_percent;
    nominalGdpGrowth[i] = bv.nominal_gdp_growth_percent;
    nominalGdp[i] = bv.nominal_gdp;
    realGdp[i] = bv.real_gdp;
    employmentGrowth[i] = bv.employment_growth;

    revenue[i] = fl.revenue;
    revenuePct[i] = fl.revenue_percent_gdp;
    primaryExp[i] = fl.primary_expenditure;
    primaryExpPct[i] = fl.primary_expenditure_percent_gdp;
    primaryBal[i] = fl.primary_balance;
    primaryBalPct[i] = fl.primary_balance_percent_gdp;
    interestExp[i] = fl.interest_expenditure;
    interestExpPct[i] = fl.interest_expenditure_percent_gdp;
    totalExp[i] = fl.total_expenditure;
    overallBal[i] = fl.overall_balance;
    overallBalPct[i] = fl.overall_balance_percent_gdp;
    debtToGdp[i] = fl.debt_to_gdp;
    debt[i] = fl.debt;

    // DSPB from 2010 onward.
    if (i > 0) {
      const nomRate = mustGet(interestLookup, year, 'nominal_interest_rate');
      const gdpG = nominalGdpGrowth[i]!;
      dspb[i] = ((debtToGdp[i - 1]! * (nomRate - gdpG)) / 100) / (1 + gdpG / 100);
    }
  }

  // --- Projection period: recursive, explicit t-1 loop ---
  for (let i = 0; i < n; i += 1) {
    const year = years[i]!;
    if (year <= weoMaxYear) continue;

    const bv = mustGet(bv1Lookup, year, 'baseline_v1');
    const fl = mustGet(fiscalLookup, year, 'fiscal baseline');
    const nomRate = mustGet(interestLookup, year, 'nominal_interest_rate');
    const variation = cvLookup.get(year) ?? 0.0;
    const inflation = bv.gdp_deflator_growth_percent;

    // Phase 1: shock productivity; employment is unchanged from baseline.
    employmentGrowth[i] = bv.employment_growth;
    labourProdGrowth[i] = bv.labour_productivity_growth + variation;

    // Phase 2: recompute GDP recursively.
    realGdpGrowth[i] =
      (1 + employmentGrowth[i]! / 100) * (1 + labourProdGrowth[i]! / 100) * 100 - 100;
    nominalGdpGrowth[i] =
      (1 + realGdpGrowth[i]! / 100) * (1 + inflation / 100) * 100 - 100;
    realGdp[i] = realGdp[i - 1]! * (1 + realGdpGrowth[i]! / 100);
    nominalGdp[i] = nominalGdp[i - 1]! * (1 + nominalGdpGrowth[i]! / 100);

    // Phase 3: expenditure recalibration. rigidity=1.0 holds the baseline LEVEL
    // (sticky, worst case); rigidity=0.0 holds the baseline SHARE of GDP.
    const baselinePexp = fl.primary_expenditure;
    const baselinePexpPct = fl.primary_expenditure_percent_gdp;
    const primaryExpWithBaselinePct = (baselinePexpPct * nominalGdp[i]!) / 100;
    const recalibration = baselinePexp - primaryExpWithBaselinePct;
    primaryExp[i] = baselinePexp - (1 - expenditureRigidity) * recalibration;

    // Phase 4: revenue holds the baseline ratio to GDP.
    revenuePct[i] = fl.revenue_percent_gdp;
    revenue[i] = (revenuePct[i]! / 100) * nominalGdp[i]!;

    if (dataRisk) {
      const revRisk = riskRevLookup.get(year) ?? 0.0;
      const expRisk = riskExpLookup.get(year) ?? 0.0;
      revenue[i] = revenue[i]! + (revRisk / 100) * nominalGdp[i]!;
      primaryExp[i] = primaryExp[i]! + (expRisk / 100) * nominalGdp[i]!;
      revenuePct[i] = (revenue[i]! / nominalGdp[i]!) * 100;
    }

    // Phase 5: recursive fiscal block — NO debt floor here (domain rule #3).
    primaryExpPct[i] = (primaryExp[i]! / nominalGdp[i]!) * 100;
    primaryBal[i] = revenue[i]! - primaryExp[i]!;
    primaryBalPct[i] = (primaryBal[i]! / nominalGdp[i]!) * 100;

    debtToGdp[i] =
      (debtToGdp[i - 1]! * (1 + nomRate / 100)) / (1 + nominalGdpGrowth[i]! / 100) -
      primaryBalPct[i]!;

    debt[i] = (debtToGdp[i]! / 100) * nominalGdp[i]!;

    interestExp[i] = (debt[i - 1]! * nomRate) / 100;
    interestExpPct[i] = (interestExp[i]! / nominalGdp[i]!) * 100;

    totalExp[i] = primaryExp[i]! + interestExp[i]!;
    overallBal[i] = revenue[i]! - totalExp[i]!;
    overallBalPct[i] = (overallBal[i]! / nominalGdp[i]!) * 100;

    dspb[i] =
      ((debtToGdp[i - 1]! * (nomRate - nominalGdpGrowth[i]!)) / 100) /
      (1 + nominalGdpGrowth[i]! / 100);
  }

  return years.map((year, i) => ({
    years: year,
    revenue: revenue[i]!,
    revenue_percent_gdp: revenuePct[i]!,
    primary_expenditure: primaryExp[i]!,
    primary_expenditure_percent_gdp: primaryExpPct[i]!,
    primary_balance: primaryBal[i]!,
    primary_balance_percent_gdp: primaryBalPct[i]!,
    interest_expenditure: interestExp[i]!,
    interest_expenditure_percent_gdp: interestExpPct[i]!,
    total_expenditure: totalExp[i]!,
    overall_balance: overallBal[i]!,
    overall_balance_percent_gdp: overallBalPct[i]!,
    debt_to_gdp: debtToGdp[i]!,
    debt: debt[i]!,
    debt_stabilizing_primary_balance: dspb[i]!,
    labour_productivity_growth: labourProdGrowth[i]!,
    real_gdp_growth_percent: realGdpGrowth[i]!,
    nominal_gdp_growth_percent: nominalGdpGrowth[i]!,
    nominal_gdp: nominalGdp[i]!,
    real_gdp: realGdp[i]!,
    employment_growth: employmentGrowth[i]!,
  }));
}
